src/modules/wind_estimator/wind_estimator_main.cpp

```diff
@@ -154,104 +206,295 @@ WindEstimatorModule::task_spawn(int argc, char *argv[])
+	}
+
+	if (_vtol_vehicle_status_sub.update(&_vtol_vehicle_status)) {
+		// _vehicle_attitude_valid = (_time_now_usec - vehicle_attitude.timestamp < 1_s) && (vehicle_attitude.timestamp > 0);
+	}
+}
+
+void AirspeedModule::update_wind_estimator_sideslip()
+{
+	bool att_valid = true;
+	const hrt_abstime time_now_usec = hrt_absolute_time();
+
+	// update wind and airspeed estimator
+	_wind_estimator_sideslip.update(time_now_usec);
+
+	if (_vehicle_local_position_valid && att_valid) {
+		Vector3f vI(_vehicle_local_position.vx, _vehicle_local_position.vy, _vehicle_local_position.vz);
+		Quatf q(_vehicle_attitude.q);
+
+		// sideslip fusion
+		_wind_estimator_sideslip.fuse_beta(time_now_usec, vI, q);
+	}
+
+	// fill message for publishing later
+	_wind_estimate_sideslip.timestamp = time_now_usec;
+	float wind[2];
+	_wind_estimator_sideslip.get_wind(wind);
+	_wind_estimate_sideslip.windspeed_north = wind[0];
+	_wind_estimate_sideslip.windspeed_east = wind[1];
+	float wind_cov[2];
+	_wind_estimator_sideslip.get_wind_var(wind_cov);
+	_wind_estimate_sideslip.variance_north = wind_cov[0];
+	_wind_estimate_sideslip.variance_east = wind_cov[1];
+	_wind_estimate_sideslip.tas_innov = _wind_estimator_sideslip.get_tas_innov();
+	_wind_estimate_sideslip.tas_innov_var = _wind_estimator_sideslip.get_tas_innov_var();
+	_wind_estimate_sideslip.beta_innov = _wind_estimator_sideslip.get_beta_innov();
+	_wind_estimate_sideslip.beta_innov_var = _wind_estimator_sideslip.get_beta_innov_var();
+	_wind_estimate_sideslip.tas_scale = _wind_estimator_sideslip.get_tas_scale();
+}
+
+void AirspeedModule::update_ground_minus_wind_airspeed()
+{
+	// calculate airspeed estimate based on groundspeed-windspeed to use as fallback
+	float TAS_north = _vehicle_local_position.vx - _wind_estimate_sideslip.windspeed_north;
+	float TAS_east = _vehicle_local_position.vy - _wind_estimate_sideslip.windspeed_east;
+	_ground_minus_wind_TAS = sqrtf(TAS_north * TAS_north + TAS_east * TAS_east);
+	_ground_minus_wind_EAS = calc_EAS_from_TAS(_ground_minus_wind_TAS, _vehicle_air_data.baro_pressure_pa,
+				 _vehicle_air_data.baro_temp_celcius);
 }
 
-int WindEstimatorModule::custom_command(int argc, char *argv[])
+
+void AirspeedModule::select_airspeed_and_publish()
+{
+	/* airspeed index:
+	/  0: first airspeed sensor valid
+	/  1: second airspeed sensor valid
+	/ -1: airspeed sensor(s) invalid, but groundspeed-windspeed estimate valid
+	/ -2: airspeed invalid (sensors and groundspeed-windspeed estimate invalid)
+	*/
+	bool find_new_valid_index = false;
+
+	// find new valid index if airspeed currently invalid (but we have sensors)
+	if ((_number_of_airspeed_sensors > 0 && _prev_airspeed_index < 0) ||
+	    (_prev_airspeed_index >= 0 && !_airspeed_validator[_prev_airspeed_index].get_airspeed_valid()) ||
+	    _prev_airspeed_index == -2) {
+
+		find_new_valid_index = true;
+	}
+
+	if (find_new_valid_index) {
+		for (int i = 0; i < _number_of_airspeed_sensors; i++) {
+			_valid_airspeed_index = -1;
```

Comment:
@sfuhrer You need to move this outside of the loop. Otherwise if you have no airspeed sensors then the index -1 is not possible.